### ISSUE TITLE
Feature: AMQP-1.0 client — `tinox.core.amqp10`

### ISSUE THREAD

**Issue by subnix-work** [subnix-work/tinox#69]
**Status: IMPLEMENTED (2026-07-24), v1.** Roadmap/status in `tasklist.md`. A pure **client** (no broker) for the AMQP-1.0 protocol (OASIS standard, e.g. ActiveMQ Artemis, Azure Service Bus, RabbitMQ 4.x natively) — a standalone module `crates/tinox-core/amqp10.tnx`, **no shared code with `amqp091.tnx`** except the same basic transport pattern (TCP dial via `httpConnFromFd`/binary-safe `httpConnReadN`/`httpConnWriteBytes`). A completely different beast from 0-9-1 at the protocol and data-model level: a generic type system with format-code prefixes instead of fixed field tags, its own frame layout, its own SASL framing layer, a three-level `Connection`→`Session`→`Link` hierarchy with credit-based flow control (`flow`/`link-credit`, MANDATORY for every transfer) instead of 0-9-1's simple `Connection`→`Channel` model with unlimited publish. Example `examples/amqp10_publish_consume.tnx`.

```tinox
import tinox.core.amqp10;

let conn = Amqp10Connection::connect("127.0.0.1", 5672, "guest", "guest");
let session = Amqp10Session::begin(conn);
var sender = Amqp10Link::attach(session, "my-sender", false, "/queues/my-queue");

var body: List<Int64> = [];
for i in 0..3 { body.push("abc".charCodeAt(i)); }
sender.publish(body, "text/plain");
sender.detach();

var receiver = Amqp10Link::attach(session, "my-receiver", true, "/queues/my-queue");
receiver.grantCredit(10);
let m = receiver.nextMessage();       // blocking pull
if m.ok { receiver.ack(m.deliveryId); }
conn.close();
```

**Architecture.** The `Amqp10Value` enum (20 variants) as a generic, recursive value codec — ONE type system for performatives, SASL frames, AND message sections, unlike 0-9-1's two separate codecs (method arguments vs. field table). `Amqp10Writer`/`Amqp10Reader` as byte builder/cursor; the writer deliberately ALWAYS uses the full 32-bit form for variable-length types (`vbin32`/`str32`/`sym32`, never the compressed 8-bit short forms) — as a result, unlike 0-9-1's `shortstr`, there is NO 255-byte write limit that could enable a Bug-71-style silent-truncation finding (empirically confirmed via `amqp10_edge_cases.tnx`, not just asserted); the reader still reads all the short forms though, since brokers are allowed to use them. `Amqp10Described` (descriptor + value) is the generic building block for performatives (`encodePerformative`/`decodePerformative`, ONE function for all 9 performative types instead of 0-9-1's method-specific `encodeMethodPayload`) AND for message sections (`properties`/`data`/`amqp-value`). `Amqp10Link::publish()` splits the body at `max-frame-size` across multiple `transfer` frames (a `more` flag instead of 0-9-1's implicit continuation via frame boundaries) and internally waits on `awaitFlow()` for credit if exhausted — credit-based flow control has no equivalent in 0-9-1 (only the coarse `basic.qos`/`prefetchCount` there).

**Hardening behavior (project convention: no silent garbage):** protocol errors (frame-header violations, unexpected performative structure) → `frameType -2`/`descriptor -1`, no silent progress; every public method returns `.errorMessage`/`.ok == false` instead of hanging or a false success. **Bug 71 (see above) is the prime example of this principle:** `nextMessage()` originally only recognized `data` body sections and returned `ok=true` with an empty body for `amqp-value` sections — fixed by recognizing both section types AND a hard `ok=false` failure when no body section is found at all (this also turns the third type not supported in v1, `amqp-sequence`, into a visible rather than silent error). **Bug 70 (see above):** `initial-delivery-count` is REQUIRED by spec when `role=Sender` — originally missing, which caused RabbitMQ 4.x to crash with `function_clause`; only discoverable via live verification against a real broker, no simulated-broker test would have reproduced the strict server-side pattern matching.

**Deliberate v1 gaps (see tasklist.md "Later"):** only one session/one link per connection (no pool), only SASL PLAIN (no `sasl-challenge`/`-response` for SCRAM & co.), delivery state only `accepted` (no `rejected`/`released`/`modified`), no transactions (`txn-id`, `declare`/`discharge`), no link recovery/resumption (`unsettled` map on `attach`), **no multi-frame reassembly on receive** (`nextMessage()` detects `more=true` and reports a hard error instead of just returning the first part — sender-side splitting AND reassembly on send are implemented and tested, just not the receive direction), no annotation-driven consumer API (the same lambda-as-method-param blocker as for 0-9-1 and WS v1), no heartbeat (an `empty` AMQP frame as keepalive)/auto-reconnect.

**Verified:** automated loopback e2e tests (a simulated broker via `spawn`/`await`, no Docker needed in `make check`): `amqp10_phase1_handshake.tnx`, `amqp10_typesystem.tnx` (30 format-code round trips + short forms + array decoding + error paths), `amqp10_frame_codec.tnx`, `amqp10_sasl_negotiation.tnx`, `amqp10_connection_session_link.tnx`, `amqp10_transfer_flow.tnx` (multi-frame publish + credit exhaustion + consume/ack round trip), `amqp10_edge_cases.tnx` (Phase 7: empty message bodies in the full publish/consume round trip, the exact multi-frame split boundary — one byte under the `max-frame-size` cutoff = 1 frame, one byte over = 2 frames, using the overhead measured at runtime via `encodeMessageBody()` instead of hard-coded byte counts — plus str8/sym8/vbin8 at the 255-byte boundary and >255-byte round trips through the writer). Every `spawn`/`await` test repeated stably 10-25x (Bug-68 caution). LIVE against real RabbitMQ (Docker, `rabbitmq:4-management` — native AMQP 1.0 in the core broker since version 4.0, no plugin needed; the old `rabbitmq:3-management` + `rabbitmq_amqp1_0` plugin combination is unsuitable as a reference, see Bug 70): a full Connect→Begin→Attach→Publish→Detach→Attach→Consume→Ack→Detach→End→Close sequence over a real queue, including manual live verification of the `amqp10_publish_consume.tnx` example program. **Cross-checked with `python-qpid-proton`** (an independent Python AMQP-1.0 client, analogous to `pika` for 0-9-1) in BOTH directions (Tinox publishes → Proton consumes, and vice versa) — confirms wire compatibility with an implementation outside our own code, and uncovered Bug 71. `make check` green.

---

---
*Imported from `bugs_fixed.md` (full history with root cause/fix/verification). Automatically closed as done.*
